// @stellar/js-xdr@4.0.0/src/config.js
'use strict';

var int = require('./int.js');
var hyper = require('./hyper.js');
var unsignedInt = require('./unsigned-int.js');
var unsignedHyper = require('./unsigned-hyper.js');
require('./xdr-type.js');
var errors = require('./errors.js');
var float = require('./float.js');
var double = require('./double.js');
var quadruple = require('./quadruple.js');
var bool = require('./bool.js');
var string = require('./string.js');
var opaque = require('./opaque.js');
var varOpaque = require('./var-opaque.js');
var array = require('./array.js');
var varArray = require('./var-array.js');
var option = require('./option.js');
var _void = require('./void.js');
var _enum = require('./enum.js');
var struct = require('./struct.js');
var union = require('./union.js');
var reference = require('./reference.js');

// eslint-disable-next-line max-classes-per-file

class SimpleReference extends reference.Reference {
  constructor(name) {
    super();
    this.name = name;
  }

  resolve(context) {
    const defn = context.definitions[this.name];
    return defn.resolve(context);
  }
}

class ArrayReference extends reference.Reference {
  constructor(childReference, length, variable = false) {
    super();
    this.childReference = childReference;
    this.length = length;
    this.variable = variable;
  }

  resolve(context) {
    let resolvedChild = this.childReference;
    let length = this.length;

    if (resolvedChild instanceof reference.Reference) {
      resolvedChild = resolvedChild.resolve(context);
    }

    if (length instanceof reference.Reference) {
      length = length.resolve(context);
    }

    if (this.variable) {
      return new varArray.VarArray(resolvedChild, length);
    }
    return new array.Array(resolvedChild, length);
  }
}

class OptionReference extends reference.Reference {
  constructor(childReference) {
    super();
    this.childReference = childReference;
    this.name = childReference.name;
  }

  resolve(context) {
    let resolvedChild = this.childReference;

    if (resolvedChild instanceof reference.Reference) {
      resolvedChild = resolvedChild.resolve(context);
    }

    return new option.Option(resolvedChild);
  }
}

class SizedReference extends reference.Reference {
  constructor(sizedType, length) {
    super();
    this.sizedType = sizedType;
    this.length = length;
  }

  resolve(context) {
    let length = this.length;

    if (length instanceof reference.Reference) {
      length = length.resolve(context);
    }

    return new this.sizedType(length);
  }
}

class Definition {
  constructor(constructor, name, cfg) {
    this.constructor = constructor;
    this.name = name;
    this.config = cfg;
  }

  // resolve calls the constructor of this definition with the provided context
  // and this definitions config values.  The definitions constructor should
  // populate the final type on `context.results`, and may refer to other
  // definitions through `context.definitions`
  resolve(context) {
    if (this.name in context.results) {
      return context.results[this.name];
    }

    return this.constructor(context, this.name, this.config);
  }
}

// let the reference resolution system do its thing
// the "constructor" for a typedef just returns the resolved value
function createTypedef(context, typeName, value) {
  if (value instanceof reference.Reference) {
    value = value.resolve(context);
  }
  context.results[typeName] = value;
  return value;
}

function createConst(context, name, value) {
  context.results[name] = value;
  return value;
}

class TypeBuilder {
  constructor(destination) {
    this._destination = destination;
    this._definitions = {};
  }

  enum(name, members) {
    const result = new Definition(_enum.Enum.create, name, members);
    this.define(name, result);
  }

  struct(name, members) {
    const result = new Definition(struct.Struct.create, name, members);
    this.define(name, result);
  }

  union(name, cfg) {
    const result = new Definition(union.Union.create, name, cfg);
    this.define(name, result);
  }

  typedef(name, cfg) {
    const result = new Definition(createTypedef, name, cfg);
    this.define(name, result);
  }

  const(name, cfg) {
    const result = new Definition(createConst, name, cfg);
    this.define(name, result);
  }

  void() {
    return _void.Void;
  }

  bool() {
    return bool.Bool;
  }

  int() {
    return int.Int;
  }

  hyper() {
    return hyper.Hyper;
  }

  uint() {
    return unsignedInt.UnsignedInt;
  }

  uhyper() {
    return unsignedHyper.UnsignedHyper;
  }

  float() {
    return float.Float;
  }

  double() {
    return double.Double;
  }

  quadruple() {
    return quadruple.Quadruple;
  }

  string(length) {
    return new SizedReference(string.String, length);
  }

  opaque(length) {
    return new SizedReference(opaque.Opaque, length);
  }

  varOpaque(length) {
    return new SizedReference(varOpaque.VarOpaque, length);
  }

  array(childType, length) {
    return new ArrayReference(childType, length);
  }

  varArray(childType, maxLength) {
    return new ArrayReference(childType, maxLength, true);
  }

  option(childType) {
    return new OptionReference(childType);
  }

  define(name, definition) {
    if (this._destination[name] === undefined) {
      this._definitions[name] = definition;
    } else {
      throw new errors.XdrDefinitionError(`${name} is already defined`);
    }
  }

  lookup(name) {
    return new SimpleReference(name);
  }

  resolve() {
    for (const defn of Object.values(this._definitions)) {
      defn.resolve({
        definitions: this._definitions,
        results: this._destination
      });
    }
  }
}

function config(fn, types = {}) {
  if (fn) {
    const builder = new TypeBuilder(types);
    fn(builder);
    builder.resolve();
  }

  return types;
}

exports.Reference = reference.Reference;
exports.config = config;
