// @stellar/js-xdr@4.0.0/src/config.js
import { Int } from './int.js';
import { Hyper } from './hyper.js';
import { UnsignedInt } from './unsigned-int.js';
import { UnsignedHyper } from './unsigned-hyper.js';
import './xdr-type.js';
import { XdrDefinitionError } from './errors.js';
import { Float } from './float.js';
import { Double } from './double.js';
import { Quadruple } from './quadruple.js';
import { Bool } from './bool.js';
import { String } from './string.js';
import { Opaque } from './opaque.js';
import { VarOpaque } from './var-opaque.js';
import { Array } from './array.js';
import { VarArray } from './var-array.js';
import { Option } from './option.js';
import { Void } from './void.js';
import { Enum } from './enum.js';
import { Struct } from './struct.js';
import { Union } from './union.js';
import { Reference } from './reference.js';

// eslint-disable-next-line max-classes-per-file

class SimpleReference extends Reference {
  constructor(name) {
    super();
    this.name = name;
  }

  resolve(context) {
    const defn = context.definitions[this.name];
    return defn.resolve(context);
  }
}

class ArrayReference extends Reference {
  constructor(childReference, length, variable = false) {
    super();
    this.childReference = childReference;
    this.length = length;
    this.variable = variable;
  }

  resolve(context) {
    let resolvedChild = this.childReference;
    let length = this.length;

    if (resolvedChild instanceof Reference) {
      resolvedChild = resolvedChild.resolve(context);
    }

    if (length instanceof Reference) {
      length = length.resolve(context);
    }

    if (this.variable) {
      return new VarArray(resolvedChild, length);
    }
    return new Array(resolvedChild, length);
  }
}

class OptionReference extends Reference {
  constructor(childReference) {
    super();
    this.childReference = childReference;
    this.name = childReference.name;
  }

  resolve(context) {
    let resolvedChild = this.childReference;

    if (resolvedChild instanceof Reference) {
      resolvedChild = resolvedChild.resolve(context);
    }

    return new Option(resolvedChild);
  }
}

class SizedReference extends Reference {
  constructor(sizedType, length) {
    super();
    this.sizedType = sizedType;
    this.length = length;
  }

  resolve(context) {
    let length = this.length;

    if (length instanceof Reference) {
      length = length.resolve(context);
    }

    return new this.sizedType(length);
  }
}

class Definition {
  constructor(constructor, name, cfg) {
    this.constructor = constructor;
    this.name = name;
    this.config = cfg;
  }

  // resolve calls the constructor of this definition with the provided context
  // and this definitions config values.  The definitions constructor should
  // populate the final type on `context.results`, and may refer to other
  // definitions through `context.definitions`
  resolve(context) {
    if (this.name in context.results) {
      return context.results[this.name];
    }

    return this.constructor(context, this.name, this.config);
  }
}

// let the reference resolution system do its thing
// the "constructor" for a typedef just returns the resolved value
function createTypedef(context, typeName, value) {
  if (value instanceof Reference) {
    value = value.resolve(context);
  }
  context.results[typeName] = value;
  return value;
}

function createConst(context, name, value) {
  context.results[name] = value;
  return value;
}

class TypeBuilder {
  constructor(destination) {
    this._destination = destination;
    this._definitions = {};
  }

  enum(name, members) {
    const result = new Definition(Enum.create, name, members);
    this.define(name, result);
  }

  struct(name, members) {
    const result = new Definition(Struct.create, name, members);
    this.define(name, result);
  }

  union(name, cfg) {
    const result = new Definition(Union.create, name, cfg);
    this.define(name, result);
  }

  typedef(name, cfg) {
    const result = new Definition(createTypedef, name, cfg);
    this.define(name, result);
  }

  const(name, cfg) {
    const result = new Definition(createConst, name, cfg);
    this.define(name, result);
  }

  void() {
    return Void;
  }

  bool() {
    return Bool;
  }

  int() {
    return Int;
  }

  hyper() {
    return Hyper;
  }

  uint() {
    return UnsignedInt;
  }

  uhyper() {
    return UnsignedHyper;
  }

  float() {
    return Float;
  }

  double() {
    return Double;
  }

  quadruple() {
    return Quadruple;
  }

  string(length) {
    return new SizedReference(String, length);
  }

  opaque(length) {
    return new SizedReference(Opaque, length);
  }

  varOpaque(length) {
    return new SizedReference(VarOpaque, length);
  }

  array(childType, length) {
    return new ArrayReference(childType, length);
  }

  varArray(childType, maxLength) {
    return new ArrayReference(childType, maxLength, true);
  }

  option(childType) {
    return new OptionReference(childType);
  }

  define(name, definition) {
    if (this._destination[name] === undefined) {
      this._definitions[name] = definition;
    } else {
      throw new XdrDefinitionError(`${name} is already defined`);
    }
  }

  lookup(name) {
    return new SimpleReference(name);
  }

  resolve() {
    for (const defn of Object.values(this._definitions)) {
      defn.resolve({
        definitions: this._definitions,
        results: this._destination
      });
    }
  }
}

function config(fn, types = {}) {
  if (fn) {
    const builder = new TypeBuilder(types);
    fn(builder);
    builder.resolve();
  }

  return types;
}

export { Reference, config };
