// @stellar/js-xdr@4.0.0/src/struct.js
'use strict';

var reference = require('./reference.js');
var xdrType = require('./xdr-type.js');
var errors = require('./errors.js');

class Struct extends xdrType.NestedXdrType {
  constructor(attributes, maxDepth) {
    const resolvedMaxDepth = maxDepth ?? new.target?._maxDepth;
    super(resolvedMaxDepth);
    this._attributes = attributes || {};
  }

  /**
   * @inheritDoc
   */
  static read(reader, remainingDepth = this._maxDepth) {
    xdrType.NestedXdrType.checkDepth(remainingDepth);

    const attributes = {};
    for (const [fieldName, type] of this._fields) {
      attributes[fieldName] = type.read(reader, remainingDepth - 1);
    }
    return new this(attributes, this._maxDepth);
  }

  /**
   * @inheritDoc
   */
  static write(value, writer) {
    if (!this.isValid(value)) {
      throw new errors.XdrWriterError(
        `${value} has struct name ${value?.constructor?.structName}, not ${
          this.structName
        }: ${JSON.stringify(value)}`
      );
    }

    for (const [fieldName, type] of this._fields) {
      const attribute = value._attributes[fieldName];
      type.write(attribute, writer);
    }
  }

  /**
   * @inheritDoc
   */
  static isValid(value) {
    return (
      value?.constructor?.structName === this.structName ||
      xdrType.isSerializableIsh(value, this)
    );
  }

  static create(
    context,
    name,
    fields,
    maxDepth = xdrType.NestedXdrType.DEFAULT_MAX_DEPTH
  ) {
    const ChildStruct = class extends Struct {};

    ChildStruct.structName = name;
    ChildStruct._maxDepth = maxDepth;
    context.results[name] = ChildStruct;

    const mappedFields = new Array(fields.length);
    for (let i = 0; i < fields.length; i++) {
      const fieldDescriptor = fields[i];
      const fieldName = fieldDescriptor[0];
      let field = fieldDescriptor[1];
      if (field instanceof reference.Reference) {
        field = field.resolve(context);
      }
      mappedFields[i] = [fieldName, field];
      // create accessors
      ChildStruct.prototype[fieldName] = createAccessorMethod(fieldName);
    }

    ChildStruct._fields = mappedFields;

    return ChildStruct;
  }
}

function createAccessorMethod(name) {
  return function readOrWriteAttribute(value) {
    if (value !== undefined) {
      this._attributes[name] = value;
    }
    return this._attributes[name];
  };
}

exports.Struct = Struct;
